Permission manager

danielb987 opened the pull request:
This PR adds a permission manager to JMRI which allows different users to have different permissions. The first step is to have permissions to load and store tables and panel files, but the design is generic and can be used for everything in JMRI, as long as someone implements it.

See [LoadXmlUserAction](java/src/jmri/configurexml/LoadXmlUserAction.java) and [LoadAndStorePermissionOwner](java/src/jmri/configurexml/LoadAndStorePermissionOwner.java) for an example on how to use it.

Note: This PR is far from complete. I publish it due to the long discussion on the subject on the [JMRI users list](https://groups.io/g/jmriusers/topic/107810289#msg233276).

bobjacobsen commented:
I'd like to understand how you see this working.

Will the user and permission information be stored in a separate file? With the user preferences in its file? Or with the panel file(s)?

How will a user sign in to JMRI? How will JMRI know to require a sign-on?  Will the user experience be the same as now if the permission system is not in use?

Can you change the logged-in user within a JMRI run, e.g. can you logout and login somebody else?

Thanks for starting on this!

danielb987 commented:
> Will the user and permission information be stored in a separate file?

Yes, in the file `PREFERENCES / .permissions.xml`. On my Linux computer, it's `~/.jmri/.permissions.xml`.

This file contains all the usernames, their passwords encoded with a salted MD5 and the users permissions.

Once the permissions are set up, it's possible for the advanced user to protect this file if desired. On Linux, the user could set it owned by root and only readonly permission for JMRI.

> Will the user experience be the same as now if the permission system is not in use?

The permission manager is disabled by default and all users has full access to everything.

> How will a user sign in to JMRI?
> Can you change the logged-in user within a JMRI run, e.g. can you logout and login somebody else?

When JMRI starts, the `Guest` user is the current user. The user can then log in by selecting the menu item `File -> Login` and enter username and password. When he is done, he can logout by selecting the menu item `File -> Logout`.

We might later add an optional login prompt when JMRI starts, before the profile dialog is shown. I initialize the profile manager early so it's prepared for that. But I don't plan to implement that now, unless someone desires it.

> How will JMRI know to require a sign-on?

Let's assume we want to protect the JMRI tables (turnouts, sensors, ...). First we create a class that implements `PermissionOwner`. It uses a ServiceProvider to notify about its existence.

The PermissionOwner registers its permissions in the permission manager. Each permission is a class that implements `Permission`. The permission owner is like a category. It groups the permissions that are related.

Let's say we have the permissions:
- View table
- Edit table
- Control bean (throw turnout, activate sensor, ...)

Then we add code to the bean tables that calls `PermissionManager.checkPermission(permission)`. If the permission manager is disabled or if the user has the permission, `checkPermission()` returns true and we can continue. But if the user doesn't have the permission, `checkPermission()` will either show a message box `Permission denied` or log a message to the log (depending on Headless or not), and return false. In this case, we abort what we where doing.

The LoadXmlAction has this code:
```
public void actionPerformed(ActionEvent e) {
    if (! InstanceManager.getDefault(PermissionManager.class)
            .checkPermission(LoadAndStorePermissionOwner.LOAD_XML_FILE_PERMISSION)) {
        return;
    }
    JFileChooser userFileChooser = getUserFileChooser();
    userFileChooser.setDialogType(javax.swing.JFileChooser.OPEN_DIALOG);
    ...
```

bobjacobsen commented:
Thank you!  That was very helpful.

dsand47 commented:
I would not expect a "user" to have more than one user type.  Maybe a radio button group?

![admin](https://github.com/user-attachments/assets/ebbea631-a4a4-4eb5-91f7-5a9ee86cb7a9)

danielb987 commented:
Thanks for the review.

> I would not expect a "user" to have more than one user type. Maybe a radio button group?

It's intentional that the same user can have different roles. You can have one role for changing panels (load and store). You can have another role for changing preferences. You can have a third role for managing other users roster. And so on. The benefit with this is that you can set up different roles for different tasks in JMRI. And then you can add and remove that role to/from users when users need to work with a particular task.

For now, there are only a couple of permissions in the permission manager. But as time passes, the number of permissions might grow. And if it does, it's useful to separate users and permissions. Define a role for a particular task, for example managing the roster, and give that role the permissions the role needs for that task. Then give that role to the users that needs it. If the role needs different permissions, you change the role and all the users that have that role will get those permissions. And if a user no longer needs those permissions, just remove the role from the users.

dsand47 commented:
I see.  The list below the comment field is a list of available roles and indicates which ones are assigned.  There are three standard roles and any number of additional roles.

It would be clearer if there was a "Roles" label above the role list.  I assumed the first three were the user type.

bobjacobsen commented:
Thanks for working on this.

I just picked up the most recent version of this branch.  I've enabled the permissions manager, but I don't see Login and Logout items in the Files menu.  Do I need to do something else to get those?

danielb987 commented:
> It would be clearer if there was a "Roles" label above the role list. I assumed the first three were the user type.

Thanks. Fixed.

danielb987 commented:
> I just picked up the most recent version of this branch. I've enabled the permissions manager, but I don't see Login and Logout items in the Files menu. Do I need to do something else to get those?

Thank you!

I hadn't implemented it yet. I have focused on the backbone and took the user interface last. But now it's almost complete, I hope. You can login and logout, change password, add new users and new roles. But remove role and remove user is still on the TODO list.

I recommend that you delete the `.permissions.xml` file. JMRI will then create a new file with users `Guest` and `Admin`. The Admin user has by default the password `jmri`. Both usernames and passwords are case sensitive. Note that you will need to restart JMRI once you have enabled the permission manager.

dsand47 commented:
The permissions menu group should be under the load/store/history group.  Adding it above messes with mouse movement memory.

dsand47 commented:
I think the DecoderPro menu is at `xml/config/parts/jmri/jmrit/roster/swing/RosterFrameMenu.xml`

----

Should the permissions menu items be hidden when the permission system is not active.  One GUI theory is that menu items should not appear and disappear.  On the other hand, the permission system will rarely be used.

danielb987 commented:
> Should the permissions menu items be hidden when the permission system is not active. One GUI theory is that menu items should not appear and disappear. On the other hand, the permission system will rarely be used.

I have the same thoughts. Since most users will never use it, I think it's reasonable that it disappears. And once you enable the permission system, it will always stay there. So it's not that it appears, disappears and appears again.

danielb987 commented:
> I think the DecoderPro menu is at `xml/config/parts/jmri/jmrit/roster/swing/RosterFrameMenu.xml`

Thanks. I tried to start DecoderPro without doing any additional change and it works fine. So it seems that its the same `File` menu for both PanelPro and DecoderPro.

dsand47 commented:
> > I think the DecoderPro menu is at `xml/config/parts/jmri/jmrit/roster/swing/RosterFrameMenu.xml`
> 
> Thanks. I tried to start DecoderPro without doing any additional change and it works fine. So it seems that its the same `File` menu for both PanelPro and DecoderPro.

Are you sure.  When running a DecoderPro compile, there are no permission menu items in the File menu.

![Screenshot 2024-08-13 at 8 45 21 PM](https://github.com/user-attachments/assets/34b7c22d-5d7d-4c3e-9aba-34e60f536da9)

 Attempting to display preferences results in "User Guest has not permission Edit preferences".

danielb987 commented:
Thank you! The problem was on my side. When I start NetBeans I need to select the main class and I thought that the main class for DecoderPro was `apps.DecoderPro.DecoderPro`. But instead I need to select `apps.gui3.dp3.DecoderPro3`.

I have now added the actions to the DecoderPro menu xml file. But the menu choices are disabled and I don't understand how to enable them.

dsand47 commented:
> I have now added the actions to the DecoderPro menu xml file. But the menu choices are disabled and I don't understand how to enable them.

These message might be relevant.
```
21:25:30,574 jmri.util.swing.GuiUtilBase           WARN  - Did not find suitable ctor for jmri.jmrit.permission.swing.LoginAction without icon [main]
21:25:30,575 jmri.util.swing.GuiUtilBase           WARN  - Did not find suitable ctor for jmri.jmrit.permission.swing.LogoutAction without icon [main]
21:25:30,575 jmri.util.swing.GuiUtilBase           WARN  - Did not find suitable ctor for jmri.jmrit.permission.swing.ChangePasswordAction without icon [main]
```

dsand47 commented:
If you want, I have done some changes to the help page.  The txt file is a diff.  The PDF is the printed web page.

[help-diff.txt](https://github.com/user-attachments/files/16607617/help-diff.txt)

[JMRI: Permissions.pdf](https://github.com/user-attachments/files/16607619/JMRI.Permissions.pdf)

danielb987 commented:
> These message might be relevant.

Thank you! I found the problem. The actions should extend `JmriAbstractAction` instead of `AbstractAction`.

I have also fixed so that DecoderPro enables/disables the Login and Logout menu items depending on whenever the user is logged in or not. But I don't see a way to hide the menu items in DecoderPro. Therefore I have decided to always show the menu items in both DecoderPro and PanelPro to make it consistent between the applications. So the Login/Logout menu items are now shown even if the permission system is disabled.

danielb987 commented:
> If you want, I have done some changes to the help page. The txt file is a diff. The PDF is the printed web page.

Thank you! How do I merge your diff? Which command should I run at the command line?

dsand47 commented:
> Thank you! How do I merge your diff? Which command should I run at the command line?

git apply ~/Desktop/help-diff.txt

Use the current path to the txt file.

Note:  I noticed some property file text changes, but I don't have time tonight.

danielb987 commented:
> Note: I noticed some property file text changes, but I don't have time tonight.

No problem. I'm going to bed now and it's not urgent to merge this PR.

dsand47 reviewed (changes requested):
The comments relate to changes to the wording of labels and class descriptions.

danielb987 commented on java/src/jmri/jmrit/permission/swing/ChangePasswordAction.java:
```diff
@@ -0,0 +1,64 @@
+
+
+/**
+ * Let a user login to the permission manager.
```
Thanks. I changed it to:
Let a user change his own password.

danielb987 commented on java/src/jmri/jmrit/permission/swing/ChangePasswordDialog.java:
```diff
@@ -0,0 +1,131 @@
+import jmri.util.swing.JmriJOptionPane;
+
+/**
+ * Dialog to add user
```
Thanks. I changed it to:
Dialog to change the user's own password.

danielb987 commented:
> The comments relate to changes to the wording of labels and class descriptions.

Thank you! All are now fixed.

dsand47 commented on java/src/jmri/NamedBeanBundle.properties:
```diff
@@ -540,7 +540,7 @@ PermissionsSystemAdmin_PermissionEditPreferences    = Edit preferences
 
-PermissionManager_UserAlreadyExistsException    = User already exists
-PermissionManager_UserDoesNotExistException     = User does not exists
+PermissionManager_UserAlreadyExistsException    = User already exist
```
This one should still be `exists`

danielb987 replied:
Thanks. Now I hope I have got it right.

dsand47 reviewed (changes requested):
Just one small change.

danielb987 commented:
I have updated the PR so that it's now possible to remove users and roles.

Two questions:
- The `Admin` user has uppercase `A`. Should it be lowercase `a` instead? `admin` instead of `Admin`?
- The `Admin` username is defined in the bundle. Should it be hardcoded in JMRI instead? I see both good and bad with this solution.

dsand47 commented:
The load permission is not being checked for load actions assigned to a PanelPro button.

[load-diff.txt](https://github.com/user-attachments/files/16628997/load-diff.txt)

bobjacobsen commented:
> I have also fixed so that DecoderPro enables/disables the Login and Logout menu items depending on whenever the user is logged in or not. But I don't see a way to hide the menu items in DecoderPro. Therefore I have decided to always show the menu items in both DecoderPro and PanelPro to make it consistent between the applications. So the Login/Logout menu items are now shown even if the permission system is disabled.

Does the permission manager only take effect after a restart?

If so, I think I can see how to make the DP menu system check for the existence (or activity?) of the permission manager before creating the login and logout menu items.  Not so much to hide them when it’s inactive, but not create them at all when the PM is inactive at startup.

Would that be an OK addition?

danielb987 commented:
@bobjacobsen 
I think it would be very good if the login/logout menu items are hidden if the permission manager is disabled. The only reason it isn't that way is because I didn't know how to solve it for DecoderPro. If you have a solution for it, I'm very happy to apply it.

danielb987 commented:
> The load permission is not being checked for load actions assigned to a PanelPro button.
> 
> [load-diff.txt](https://github.com/user-attachments/files/16628997/load-diff.txt)

Thank you! I have added it now. I wasn't aware of that feature. It's a lot of features of JMRI that I have never used.

danielb987 commented:
> I think it would be very good if the login/logout menu items are hidden

I mean not created at all.

KenC57 commented:
Bob J,

Other option if people really thought those disabled entries
(Login/Logout) were a bad thing, would be having the whole permission
system require setting something in the Preferences somewhere. If you
didn't set it active or configure some part of the permissions, it
wouldn't be displayed. At least that would mean for all current users,
they don't see it until they say they want it via the Preferences. I
know some get bent out of shape when they see somethings that are
disabled. Kind of like a package marked "Don't open until Xmas".

-ken c

danielb987 commented:
> Other option if people really thought those disabled entries (Login/Logout) were a bad thing, would be having the whole permission system require setting something in the Preferences somewhere.

There is already that setting. See the first checkbox `Enable permission manager`. It's disabled by default. The user must select it to use the permission manager at all.

![PermissionManager](https://github.com/user-attachments/assets/8dd839e5-3b23-4937-8812-1258a22f63b2)

dsand47 commented:
I created a role that allows loading xml file but not storing.  It creates permission denied errors.

[permissions.txt](https://github.com/user-attachments/files/16630861/permissions.txt)

danielb987 commented:
> I created a role that allows loading xml file but not storing. It creates permission denied errors.

Thanks for finding that! Fixed.

I have also made the OK button the default button in the login dialog and fixed a couple of places in the preference dialog which didn't set the dirty flag.

danielb987 commented:
I have changed encryption from MD5 to SHA-256 by request of Robert Middleton and @dsand47.

This breaks all the current passwords so you will either have to delete the `.permissions.xml` file or change `<Enabled>yes</Enabled>` to ` <Enabled>no</Enabled>` and then change the passwords in the preference panel.

dsand47 commented:
The attached diff shows how I "think" you can filter the DecoderPro menu entries.  

[dp-diff.txt](https://github.com/user-attachments/files/16633313/dp-diff.txt)

danielb987 commented:
> The attached diff shows how I "think" you can filter the DecoderPro menu entries.

Thanks. But I don't follow how you are thinking. The loading of the menu items should depend on `InstanceManager.getDefault(jmri.PermissionManager.class).isEnabled()` or `jmri.jmrit.permission.DefaultPermissionManager.isEnabled()`.

dsand47 commented:
I think you can put the conditional menu logic where I have the log statement.

This diff is a reference, not a solution. 


----- Original message -----
From: Daniel ***@***.***>
To: JMRI/JMRI ***@***.***>
Cc: Dave Sand ***@***.***>, Mention ***@***.***>
Subject: Re: [JMRI/JMRI] Permission manager (PR #13317)
Date: Friday, August 16, 2024 1:04 AM



> The attached diff shows how I "think" you can filter the DecoderPro menu entries.
> 
Thanks. But I don't follow how you are thinking. The loading of the menu items should depend on `InstanceManager.getDefault(jmri.PermissionManager.class).isEnabled()` or `jmri.jmrit.permission.DefaultPermissionManager.isEnabled()`.


—
Reply to this email directly, view it on GitHub <https://github.com/JMRI/JMRI/pull/13317#issuecomment-2292874400>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AED27YDZMQOGAWZYX6DCZRTZRWI6LAVCNFSM6AAAAABMJHJKA6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEOJSHA3TINBQGA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

danielb987 commented:
I have got it to work now.

danielb987 commented:
I have now updated the PR so that the Login/Logout/ChangePassword menu items are not loaded if the permission manager is not active.

RosterFrameMenu.xml
```
    <node name="Login">
        <adapter>jmri.jmrit.permission.swing.LoginAction</adapter>
        <instanceManagerOptionalClass>jmri.PermissionManager</instanceManagerOptionalClass>
        <optionalFunction>isEnabled</optionalFunction>
    </node>
```
I have two optional elements:
- `instanceManagerOptionalClass` - it gives the class name for a class to be loaded by the InstanceManager
- `optionalFunction` - it gives the name of a method without any parameters that returns a boolean value. If the value is `true`, the menu item is loaded.

I have also changed the permission preferences so that the user is requested to restart JMRI when he changes the permissions.

@bobjacobsen 
Is `RosterFrameMenu.xml` documented anywhere? Documentation or tests? Should I add a note about this somewhere?

danielb987 commented:
> @bobjacobsen
> Is `RosterFrameMenu.xml` documented anywhere? Documentation or tests? Should I add a note about this somewhere?

I think I found it. `xml/schema/guinode.xsd`

bobjacobsen commented:
> Is RosterFrameMenu.xml documented anywhere? Documentation or tests? Should I add a note about this somewhere?

There's a mention of the ability to create menus from XML on the [developer's Swing page](https://www.jmri.org/help/en/html/doc/Technical/Swing.shtml), but it's not very specific.  It refers to the JavaDoc for some classes, but they don't say very much.

The history here is that when DecoderPro was being reimplemented as DecoderPro3 (DecoderPro 1 was a table driven interface that didn't scale as well as the tabbed one of DecoderPro2 and 3), there was some effort toward a general Swing-handling approach that would eventually be used for other things.  A lot of effort went into that, and it works pretty well.  But there was never a really neat GUI for PanelPro developed that would require it, so the various PanelPro interfaces basically stayed as they are, with multiple independent windows.  I think the only exception is the Throttle window.

Most of the advanced features of the DP3 main roster window aren't even used, I suspect:  Does anybody tear off the tool bar at the top and reposition it?

I'm not sure whether this would benefit from more documentation or not.  it's not really an active area of use, so the effort to write it up might not pay off.  But it's not likely to get any use unless it's better documented....

bobjacobsen commented:
> The Admin user has uppercase A. Should it be lowercase a instead? admin instead of Admin?

I have a small preference for the lower-case 'admin'.  AFAIK, user names on Mac and Linux are traditionally all lower case. Not sure about Windows, and I'm sure there are people out there with Initialuppercase user names, so it's not a big deal.

> The Admin username is defined in the bundle. Should it be hardcoded in JMRI instead? I see both good and bad with this solution.

Perhaps I haven't thought of the "bad" part, but this seems good to me.  If somebody really needs to override it, they can provide a local update to the bundle.

dsand47 reviewed (approved):
It is looking really good.  The menu changes work as expected.  The "restart" on Save is a good idea.

danielb987 commented:
> I have a small preference for the lower-case 'admin'. AFAIK, user names on Mac and Linux are traditionally all lower case.

I agree. It's changed by PR #13334.

danielb987 commented:
#13335 adds the release note. I did this in a separate PR since I wanted to include the link to documentation and I wanted to check the link before adding the release note.

That PR also includes the missing sidebar. The menu on the left was missing on the HTML page for the permission manager.
